README.md
# ext-solrfeedimport
This TYPO3 CMS Extensions helps you import XML &amp; RSS Feed into TYPO3 CMS Solr 
It was written for TYPO3 CMS 4.5. by the Apache Solr Team at dkd Internet Service GmbH http://www.dkd.de/

The following libraries are included

* MagpieRSS: a simple RSS integration tool from http://magpierss.sourceforge.net License GPL

Please feel free to contribute by creating a pull request.


ChangeLog
2010-08-18  Ingo Renner  <ingo.renner@dkd.de>

	* Added commits after deletion and indexing of feed items
	* Fixed encoding of non-UTF-8 feeds to use the proper TYPO3 API
	* Fixed doubled type in Solr document ID
	* Fixed creation of Solr documents from feed items, group field was missing

2010-08-04  Markus Goldbach  <markus.goldbach@dkd.de>

	* Added Documentation

2010-07-28  Markus Goldbach  <markus.goldbach@dkd.de>

	* Fixed bugs after extension cleanup
	* Added a check if the feed is UTF-8

2010-07-27  Makrus Goldbach  <markus.goldbach@dkd.de>

	* Added language support for rss items
	* Added language support for solr import
	* Changed Extensionstatus from alpha to beta

2010-07-14  Markus Goldbach  <markus.goldbach@dkd.de>

	* Added a function to remove all rss items to solr before index them

2010-07-13  Markus Goldbach  <markus.goldbach@dkd.de>

	* Initial import


tca.php
<?php
if (!defined ('TYPO3_MODE')) 	die ('Access denied.');

$TCA['tx_solrfeedimport_feed'] = array(
	'ctrl' => $TCA['tx_solrfeedimport_feed']['ctrl'],
	'interface' => array(
		'showRecordFieldList' => 'sys_language_uid,l18n_parent,l18n_diffsource,hidden,url,description'
	),
	'feInterface' => $TCA['tx_solrfeedimport_feed']['feInterface'],
	'columns' => array(
		'sys_language_uid' => array(
			'exclude' => 1,
			'label'  => 'LLL:EXT:lang/locallang_general.xml:LGL.language',
			'config' => array(
				'type'                => 'select',
				'foreign_table'       => 'sys_language',
				'foreign_table_where' => 'ORDER BY sys_language.title',
				'items' => array(
					array('LLL:EXT:lang/locallang_general.xml:LGL.allLanguages', -1),
					array('LLL:EXT:lang/locallang_general.xml:LGL.default_value', 0)
				)
			)
		),
		'l18n_parent' => array(
			'displayCond' => 'FIELD:sys_language_uid:>:0',
			'exclude'     => 1,
			'label'       => 'LLL:EXT:lang/locallang_general.xml:LGL.l18n_parent',
			'config'      => array(
				'type'  => 'select',
				'items' => array(
					array('', 0),
				),
				'foreign_table'       => 'tx_solrfeedimport_feed',
				'foreign_table_where' => 'AND tx_solrfeedimport_feed.pid=###CURRENT_PID### AND tx_solrfeedimport_feed.sys_language_uid IN (-1,0)',
			)
		),
		'l18n_diffsource' => array(
			'config' => array(
				'type' => 'passthrough'
			)
		),
		'hidden' => array(
			'exclude' => 1,
			'label'   => 'LLL:EXT:lang/locallang_general.xml:LGL.hidden',
			'config'  => array(
				'type'    => 'check',
				'default' => '0'
			)
		),
		'url' => array(
			'exclude' => 1,
			'label' => 'LLL:EXT:solr_feedimport/locallang_db.xml:tx_solrfeedimport_feed.url',
			'config' => array(
				'type'     => 'input',
				'size'     => '15',
				'max'      => '255',
				'checkbox' => '',
				'eval'     => 'trim',
				'wizards'  => array(
					'_PADDING' => 2,
					'link'     => array(
						'type'         => 'popup',
						'title'        => 'Link',
						'icon'         => 'link_popup.gif',
						'script'       => 'browse_links.php?mode=wizard',
						'JSopenParams' => 'height=300,width=500,status=0,menubar=0,scrollbars=1'
					)
				)
			)
		),
		'description' => array(
			'exclude' => 1,
			'label' => 'LLL:EXT:solr_feedimport/locallang_db.xml:tx_solrfeedimport_feed.description',
			'config' => array(
				'type' => 'input',
				'size' => '30',
				'eval' => 'required',
			)
		),
	),
	'types' => array(
		'0' => array('showitem' => 'sys_language_uid;;;;1-1-1, l18n_parent, l18n_diffsource, hidden;;1, url, description')
	),
	'palettes' => array(
		'1' => array('showitem' => '')
	)
);
?>

ext_tables.php
<?php
if (!defined ('TYPO3_MODE')) 	die ('Access denied.');

$TCA['tx_solrfeedimport_feed'] = array(
	'ctrl' => array(
		'title'                    => 'LLL:EXT:solr_feedimport/locallang_db.xml:tx_solrfeedimport_feed',
		'label'                    => 'description',
		'tstamp'                   => 'tstamp',
		'crdate'                   => 'crdate',
		'cruser_id'                => 'cruser_id',
		'languageField'            => 'sys_language_uid',
		'transOrigPointerField'    => 'l18n_parent',
		'transOrigDiffSourceField' => 'l18n_diffsource',
		'default_sortby'           => 'ORDER BY crdate',
		'delete'                   => 'deleted',
		'enablecolumns'            => array(
			'disabled' => 'hidden',
		),
		'dynamicConfigFile'        => t3lib_extMgm::extPath($_EXTKEY) . 'tca.php',
		'iconfile'                 => t3lib_extMgm::extRelPath($_EXTKEY) . 'icon_tx_solrfeedimport_feed.gif',
	),
	'feInterface' => array()
);
?>

scheduler/class.tx_solrfeedimport_scheduler_feedindextask.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2010 Markus Goldbach <markus.goldbach@dkd.de>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

require_once(t3lib_extMgm::extPath('solr_feedimport') . 'lib/magpierss/rss_fetch.inc');


/**
 * Scheduler Task to read feeds and index them into Solr
 *
 * @package TYPO3
 * @subpackage solr_feedimport
 * @author Markus Goldbach <markus.goldbach@dkd.de>
 */
class tx_solrfeedimport_scheduler_FeedIndexTask extends tx_scheduler_Task {

	/**
	 * Type of the indexed Solr document
	 *
	 * @var	string
	 */
	const ITEM_TYPE = 'tx_solrfeedimport_feeditem';

	/**
	 * Solr connection
	 *
	 * @var	tx_solr_SolrService
	 */
	protected $solr = NULL;


	/**
	 * Main method to execute the scheduler task.
	 *
	 * @see	typo3/sysext/scheduler/tx_scheduler_Task::execute()
	 * @return	boolean	True if the task executed without errors, false otherwise.
	 */
	public function execute() {
		$indexedFeeds = array();
		$previousSolrConnection = NULL;

			//deactivate cache
		define('MAGPIE_CACHE_ON', FALSE);
			//TODO copy the magpie cache to typo3temp
		$feeds = $this->getFeeds();

		foreach ($feeds as $feed) {
			$this->solr = $this->getSolrConnectionByFeed($feed);

			if ($previousSolrConnection != $this->solr) {
					// clean old index documents, we need to do a full import,
					// incremental imports don't work due to missing identifiers
				$this->solr->deleteByType(self::ITEM_TYPE);
				$this->solr->commit();
			}

			$indexedFeeds[$feed['uid']] = $this->indexFeed($feed);
			$this->solr->commit();

			$previousSolrConnection = $this->solr;
		}

		return $this->feedsIndexedSuccessfully($indexedFeeds);
	}

	/**
	 * Get the solr connection for a feed record
	 *
	 * @param	array	An feed record
	 * @return	tx_solr_SolrService	The solr connection for a feed record.
	 */
	protected function getSolrConnectionByFeed(array $feedRecord) {
		try {
			$solr = t3lib_div::makeInstance('tx_solr_ConnectionManager')->getConnectionByPageId(
				$feedRecord['pid'],
				$feedRecord['sys_language_uid']
			);
		} catch (tx_solr_NoSolrConnectionFoundException $e) {
				// TODO logging
			continue;
		}

		return $solr;
	}

	/**
	 * Get all active feeds from database
	 *
	 * @return	array	All active feeds
	 */
	protected function getFeeds() {
		$feeds = $GLOBALS['TYPO3_DB']->exec_SELECTgetRows(
			'*',
			'tx_solrfeedimport_feed',
			'deleted = 0 AND hidden = 0'
		);

		return $feeds;
	}

	/**
	 * Index the content of a single feed
	 *
	 * @param	array	A feed record
	 * @return	boolean	status of indexing
	 */
	protected function indexFeed(array $feedRecord) {
		$indexed = FALSE;

		try {
			$solrDocuments = array();

			$feedContent = fetch_rss($feedRecord['url']);
			$itemIndex   = 0;
			foreach ($feedContent->items as $item) {
				$solrDocuments[] = $this->feedItemToDocument($feedRecord, $item, $itemIndex, $feedContent->encoding);
				$itemIndex++;
			}

			$response = $this->solr->addDocuments($solrDocuments);
			if ($response->getHttpStatus() == 200) {
				$indexed = TRUE;
			}
		} catch (Exception $e) {
				// TODO logging
			continue;
		}

		return $indexed;
	}

	/**
	 * Transforms a feed item into a Solr document
	 *
	 * @param	array	Feed record of the feed item
	 * @param	array	The feed element which you want to index
	 * @param	integer	Iterator number of feed element
	 * @param	string	Encodeing of feed
	 * @return	Apache_Solr_Document a solr document
	 */
	protected function feedItemToDocument(array $feed, array $feedItem, $itemIndex, $feedEncoding) {
		$document = t3lib_div::makeInstance('Apache_Solr_Document');

			// field mapping
		$document->addField('type',     self::ITEM_TYPE);
		$document->addField('appKey',   'EXT:solr_feedimport');
		$document->addField('id',       tx_solr_Util::getDocumentId(
			self::ITEM_TYPE,
			$feed['pid'],
			$feed['uid'] . '_' . $itemIndex
		));
		$document->addField('siteHash', tx_solr_Util::getSiteHash($feed['pid']));
#		$document->addField('title',    $this->getUtf8EncodedString($feedItem['title'], $feedEncoding));
#		$document->addField('content',  $this->getUtf8EncodedString(strip_tags($feedItem['description']), $feedEncoding));
		$document->addField('title',    ($feedEncoding == 'UTF-8') ? $feedItem['title'] : utf8_encode($feedItem['title']));
		$document->addField('content',  ($feedEncoding == 'UTF-8') ? strip_tags($feedItem['description']) : utf8_encode(strip_tags($feedItem['description'])));
		$document->addField('url',      $this->getLink($feedItem));
		$document->addField('group',    '0');
		$document->addField('language', $feed['sys_language_uid']);

		return $document;
	}

	/**
	 * Returns an utf-8 encoded string. If the string is not utf-8 encoded yet,
	 * it is converted to utf-8.
	 *
	 * @param	string	String to be checked and converted on demand.
	 * @param	string	Encoding of the string.
	 * @return	string	Utf-8 encoded string.
	 */
	protected function getUtf8EncodedString($string, $sourceEncoding) {
		$utf8Aliases = array('utf-8', 'utf8', 'UTF-8', 'UTF8');
		$utf8EncodedString = '';

		$charsetConverter = t3lib_div::makeInstance('t3lib_cs');

		if (in_array($sourceEncoding, $utf8Aliases)) {
			$utf8EncodedString = $string;
		} else {
			$utf8EncodedString = $charsetConverter->utf8_encode($string, $sourceEncoding);
		}

		return $utf8EncodedString;
	}

	/**
	 * Checks if all feed items have been indexed
	 *
	 * @param	array	Array of all Solr responses
	 * @return	boolean	Over all status of all indexed items
	 */
	protected function feedsIndexedSuccessfully(array $feedIndexedStatuses) {
		$indexed = TRUE;

		foreach ($feedIndexedStatuses as $feedUid => $feedIndexedStatus) {
			if (!$feedIndexedStatus) {
				$indexed = FALSE;

					// TODO might throw an exception explaining which feed failed

				break;
			}
		}

		return $indexed;
	}

	/**
	 * Gets the link for the feed item.
	 * MAGPIERSS uses link_ or link index when parsing feeds
	 *
	 * @param	array	A feed item from which you want the link
	 * @return	string	The link of a feed item
	 */
	protected function getLink(array $feedItem) {

		if (isset($feedItem['link'])) {
			$feedLink = $feedItem['link'];
		} else {
			$feedLink = $feedItem['link_'];
		}

		return $feedLink;
	}

}


if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/solr_feedimport/scheduler/class.tx_solrfeedimport_scheduler_indextask.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/solr_feedimport/scheduler/class.tx_solrfeedimport_scheduler_indextask.php']);
}

?>

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "solr_feedimport".
#
# Auto generated 27-07-2010 10:32
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Solr RSS and Atom Feed Import',
	'description' => 'Imports RSS and Atom Feeds into Solr',
	'category' => 'misc',
	'author' => 'Markus Goldbach',
	'author_email' => 'markus.goldbach@dkd.de',
	'shy' => '',
	'dependencies' => 'solr',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'beta',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.0.3',
	'constraints' => array(
		'depends' => array(
			'solr' => '1.2.0',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
);

?>